File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/filters/LanguageFilterFragment.java
package com.pavelprymak.propodcast.presentation.screens.filters;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.pavelprymak.propodcast.App;
import com.pavelprymak.propodcast.R;
import com.pavelprymak.propodcast.databinding.FragmentLanguageFilterBinding;
import com.pavelprymak.propodcast.presentation.adapters.LanguageAdapter;
import com.pavelprymak.propodcast.presentation.adapters.LanguageClickListener;
import com.pavelprymak.propodcast.presentation.viewModels.LanguageViewModel;

import static com.pavelprymak.propodcast.utils.SettingsPreferenceManager.ALL_LANGUAGES;


public class LanguageFilterFragment extends Fragment implements LanguageClickListener {
    private FragmentLanguageFilterBinding mBinding;
    private LanguageAdapter mAdapter;
    private LanguageViewModel mLanguageViewModel;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mLanguageViewModel = ViewModelProviders.of(this).get(LanguageViewModel.class);
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_language_filter, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        prepareRecycler();
        mLanguageViewModel.getLanguages().observe(this, languages -> {
            if (languages != null) {
                String selectedLanguage = App.mSettings.getFilterLanguage();
                mAdapter.updateList(languages, selectedLanguage);
                if (!selectedLanguage.equals(ALL_LANGUAGES)) {
                    int selectedItemPosition = mAdapter.getPositionByLanguageName(selectedLanguage);
                    mBinding.recyclerLanguages.smoothScrollToPosition(selectedItemPosition);
                }
            }
        });
        mBinding.toolbar.setNavigationOnClickListener(v -> {
            if (getActivity() != null)
                getActivity().onBackPressed();
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mLanguageViewModel.getLanguages().removeObservers(this);
    }

    private void prepareRecycler() {
        mAdapter = new LanguageAdapter(this);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        mBinding.recyclerLanguages.setLayoutManager(layoutManager);
        mBinding.recyclerLanguages.setHasFixedSize(true);
        mBinding.recyclerLanguages.setAdapter(mAdapter);
    }

    @Override
    public void onLanguageItemClick(String language) {
        App.mSettings.saveFilterLanguage(language);
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/filters/LanguageFilterFragment.kt
package com.pavelprymak.propodcast.presentation.screens.filters


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pavelprymak.propodcast.App
import com.pavelprymak.propodcast.MainActivity
import com.pavelprymak.propodcast.R
import com.pavelprymak.propodcast.presentation.adapters.LanguageAdapter
import com.pavelprymak.propodcast.presentation.adapters.LanguageClickListener
import com.pavelprymak.propodcast.presentation.viewModels.LanguageViewModel
import com.pavelprymak.propodcast.utils.SettingsPreferenceManager
import com.pavelprymak.propodcast.utils.SettingsPreferenceManager.ALL_LANGUAGES
import com.pavelprymak.propodcast.utils.otto.filters.EventUpdateLanguageFilter
import kotlinx.android.synthetic.main.fragment_language_filter.*
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel


class LanguageFilterFragment : Fragment(), LanguageClickListener {
    private var mAdapter: LanguageAdapter? = null
    private val mLanguageViewModel: LanguageViewModel by viewModel()
    private val mSettingsPref: SettingsPreferenceManager by inject()


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (activity is MainActivity) {
            if (!resources.getBoolean(R.bool.isTablet)) {
                (activity as MainActivity).setNavViewVisibility(false)
            }
        }
        return inflater.inflate(R.layout.fragment_language_filter, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        prepareRecycler()
        mLanguageViewModel.languages.observe(this, Observer { languages ->
            if (languages != null) {
                val selectedLanguage = mSettingsPref.filterLanguage
                mAdapter?.updateList(languages, selectedLanguage)
                if (selectedLanguage != ALL_LANGUAGES) {
                    val selectedItemPosition = mAdapter?.getPositionByLanguageName(selectedLanguage)
                    recyclerLanguages.smoothScrollToPosition(selectedItemPosition ?: 0)
                }
            }
        })
        toolbar.setNavigationOnClickListener { activity?.onBackPressed() }
        if (resources.getBoolean(R.bool.isTablet)) {
            toolbar.navigationIcon = null
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mLanguageViewModel.languages.removeObservers(this)
    }

    private fun prepareRecycler() {
        mAdapter = LanguageAdapter(this)
        val layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        recyclerLanguages.layoutManager = layoutManager
        recyclerLanguages.setHasFixedSize(true)
        recyclerLanguages.adapter = mAdapter
    }

    override fun onLanguageItemClick(language: String) {
        mSettingsPref.saveFilterLanguage(language)
        if (resources.getBoolean(R.bool.isTablet)) {
            App.eventBus.post(EventUpdateLanguageFilter())
        }
    }
}
